test/e2e/storage/testsuites/base.go

```diff
@@ -0,0 +1,330 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testsuites
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+)
+
+// TestSuite represents an interface for a set of tests whchi works with TestDriver
+type TestSuite interface {
+	// getName returns the name of this TestSuite
+	getName() string
+	// getTestPatterns returns a slice of TestPattern to be tested with this TestSuite
+	getTestPatterns() []testpatterns.TestPattern
+	// skipUnsupportedTest skips the test if this TestSuite is not suitable to be tested with the combination of TestPattern and TestDriver
+	skipUnsupportedTest(testpatterns.TestPattern, drivers.TestDriver)
+	// initTestResource initializes and returns the TestResource to be used with this TestSuite
+	initTestResource() TestResource
+	// createTestInput creates and returns TestInput that is passed to test function
+	createTestInput(testpatterns.TestPattern, TestResource) TestInput
+	// execTest executes test of the testpattern for the driver
+	execTest(drivers.TestDriver, testpatterns.TestPattern)
+}
+
+// TestResource represents an interface for resources that is used by TestSuite
+type TestResource interface {
+	// setupResource sets up test resources to be used for the tests with the
+	// combination of TestDriver and TestPattern
+	setupResource(drivers.TestDriver, testpatterns.TestPattern)
+	// cleanupResource clean up the test resources created in SetupResource
+	cleanupResource()
+}
+
+// TestInput represents an interface for an input to be passed for the test function
+type TestInput interface {
```

Comment:
this seems like a strange interface to me, could you explain why it is necessary?

Reply:
It might be O.K just to make ```prepareTest()``` to accept ```interface{}``` and do type assertion from ```interface{}```, not from ```TestInput```.
My intention was to ensure that all ```*TestInput``` is to use for ```TestInput```.
I'm new to golang and lack of this kind of knowledge. So I will fix as so, if it is better to implement this way in this case.